pkg/itemblock/itemblock.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package itemblock
+
+import (
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type ItemBlock struct {
+	Log   logrus.FieldLogger
+	Items []ItemBlockItem
+}
+
+type ItemBlockItem struct {
+	Gr           schema.GroupResource
+	Item         *unstructured.Unstructured
+	PreferredGVR schema.GroupVersionResource
+}
+
+func (ib *ItemBlock) AddUnstructured(gr schema.GroupResource, item *unstructured.Unstructured, preferredGVR schema.GroupVersionResource) {
+	ib.Items = append(ib.Items, ItemBlockItem{
+		Gr:           gr,
+		Item:         item,
+		PreferredGVR: preferredGVR,
+	})
+}
+
+// Could return multiple items if EnableAPIGroupVersions is set. The item matching the preferredGVR is returned first
+func (ib *ItemBlock) FindItem(gr schema.GroupResource, namespace, name string) []ItemBlockItem {
+	var itemList []ItemBlockItem
+	var returnList []ItemBlockItem
+
+	for _, item := range ib.Items {
+		if item.Gr == gr && item.Item != nil && item.Item.GetName() == name && item.Item.GetNamespace() == namespace {
+			itemGV, err := schema.ParseGroupVersion(item.Item.GetAPIVersion())
+			if err == nil && item.PreferredGVR.GroupVersion() == itemGV {
+				returnList = append(returnList, item)
+			} else {
```

Comment:
This logic does not look right.
The error branch will append the item even if the error was reported.

Reply:
@blackpiglet this func returns a list of items matching the GR (could be more than one if EnableAPIGroupVersions is set). The purpose of the PreferredGVR comparison is that we want to return the item corresponding to the preferred GVR first, followed by the rest. So we loop over the list and when we find an item that's the preferred GVR, we add it to returnList, else we add to itemList -- then at the end, we append itemList to preferredList. So after filtering out the non-matching items (first `if`), there are 3 possibilities:
1)  We parse preferredGVR, compare it to item GVR and it's a match, so put the item at the front of the list (if branch)
2) We parse preferredGVR, compare it to item GVR and it's *not* a match, so put the item in the "other items" list (else branch)
3) We fail to parse preferredGVR, so we just put the item in the "other items" list (else branch).

Basically if we ever fail to parse preferred GVR, then we just return the list of items in the order we have them. If we parse it *and* there's a match (there should be), then that one goes first. We really don't expect this to ever fail, since the preferred GVR string was determined by the item collector, so there's no reason it would fail parsing -- but since ParseGroupVersion has the possibility of returning an error, we need to handle that unlikely case -- but we don't want to error out if it happened, since it would be a non-fatal error. Maybe I should add a debug log here in the case that error is not nil and separate that `if` from the comparison test to make the logic clearer, though.